File: backend/controllers/category.go
package controllers

import (
	"finalBeFe/database"
	"finalBeFe/model"

	"github.com/gofiber/fiber/v2"
)

func CreateCategory(c *fiber.Ctx) error {
	var category model.Category

	if err := c.BodyParser(&category); err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "invalid request body",
		})
	}

	if category.Name == "" {
		return c.Status(400).JSON(fiber.Map{
			"error": "category name is required",
		})
	}

	if err := database.DB.Create(&category).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": err.Error(),
		})
	}

	return c.Status(201).JSON(category)
}

func GetAllCategory(c *fiber.Ctx) error {
	var categories []model.Category

	if err := database.DB.Preload("Courses").Find(&categories).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "gagal mengambil data kategori",
		})
	}

	var response []model.CategoryResponse

	for _, cat := range categories {
		var courseList []model.CourseResponse
		for _, course := range cat.Courses {
			courseList = append(courseList, model.CourseResponse{
				Title:       course.Title,
				Description: course.Description,
				Cover:       course.Cover,
				VideoURL:    course.VideoURL,
			})
		}

		response = append(response, model.CategoryResponse{
			Name:    cat.Name,
			Courses: courseList,
		})
	}

	return c.Status(200).JSON(fiber.Map{
		"category": response,
	})
}

func GetCategoryByID(c *fiber.Ctx) error {
	id := c.Params("id")
	var category model.Category

	if err := database.DB.Preload("Courses").First(&category, "id = ?", id).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{
			"error": "kategori tidak ditemukan",
		})
	}

	var courseList []model.CourseResponse
	for _, course := range category.Courses {
		courseList = append(courseList, model.CourseResponse{
			Title:       course.Title,
			Description: course.Description,
			Cover:       course.Cover,
			VideoURL:    course.VideoURL,
		})
	}

	response := model.CategoryResponse{
		Name:    category.Name,
		Courses: courseList,
	}

	return c.Status(200).JSON(response)
}

func UpdateCategory(c *fiber.Ctx) error {
	id := c.Params("id")
	var category model.Category

	if err := database.DB.First(&category, "id = ?", id).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{
			"error": "kategori tidak ditemukan",
		})
	}

	var input struct {
		Name string `json:"name"`
	}

	if err := c.BodyParser(&input); err != nil {
		return c.Status(400).JSON(fiber.Map{
			"error": "format input tidak valid",
		})
	}

	category.Name = input.Name

	if err := database.DB.Save(&category).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "gagal memperbarui kategori",
		})
	}

	return c.Status(200).JSON(fiber.Map{
		"message":  "kategori berhasil diperbarui",
		"category": category,
	})
}


func DeleteCategory(c *fiber.Ctx) error {
	id := c.Params("id")
	var category model.Category

	if err := database.DB.First(&category, "id = ?", id).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{
			"error": "kategori tidak ditemukan",
		})
	}

	if err := database.DB.Delete(&category).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{
			"error": "gagal menghapus kategori",
		})
	}

	return c.Status(200).JSON(fiber.Map{
		"message": "kategori berhasil dihapus",
	})
}

File: frontend/src/service/category.js
const API_URL = import.meta.env.VITE_API_URL;

// Categories Course
export const GetCategories = async () => {
    try {
        const res = await fetch(`${API_URL}/categories`, {
            method: "GET",
            credentials: "include",
        });
        if(!res.ok) throw new Error("Failed to fetch categories");

        const data = await res.json();
        console.log("Category response:", data);
        return data.Category || data.category || [];
    } catch (error) {
        console.error("Error fetching categories:", error);
        return[];
    }
};

// Categoris Course By Id
export const GetCategoryById = async (id) => {
    try {
        const res = await fetch(`${API_URL}/categories/${id}`, {
            credentials: "include",
        });

        if(!res.ok) throw new Error("Failed to fetch category");

        const data = await res.json();

        return{
            id: id,
            name: data.name,
            course: data.course || [],
        };
    } catch (error) {
        console.error("Get category error:", error);
        return null;
    }
};

// Create Category Course
export const CreateCategory = async (name) => {
    try {
        const res = await fetch(`${API_URL}/categories`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
            },
            credentials: "include",
            body: JSON.stringify({ name }),
        });

        const data = await res.json();

        if(!res.ok) throw new Error(data.error || "Failed to create category");

        return data;
    } catch (error) {
        console.error("Create category error:", error);
        throw error;
    }
};

// Update Category Course
export const UpdateCategory = async (id, name) => {
    try {
        const res = await fetch(`${API_URL}/categories/${id}`, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json",
            },
            credentials: "include",
            body: JSON.stringify({ name }),
        });

        const data = await res.json();

        if(!res.ok) throw new Error(data.error || "Failed to update category");

        return data;
    } catch (error) {
        console.error("Update category error:", error);
        throw error;
    }
};

// Delete Category Course
export const DeleteCategory = async (id) => {
    try {
        const res = await fetch(`${API_URL}/categories/${id}`, {
            method: "DELETE",
            credentials: "include",
        });

        const data = await res.json();

        if(!res.ok) throw new Error(data.error || "Failed to delete category");

        return data;
    } catch (error) {
        console.error("Delete category error:", error);
        throw error;
    }
};